### PR TITLE
Replace manual mmap with llvm::MemoryBuffer

### LINKED ISSUE #840: Replace Fortran::parser::SourceFile::ReadFile with llvm::sys::fs::readNativeFile
LLVM's implementation is portable to systems like Windows.

### PR DESCRIPTION
**Pull request by DavidTruby**
Fixes #840

### CODE REVIEW

**Comment by DavidTruby** [comment]
@sscalpone could you check if this has the defective fd behaviour you mentioned on the mailing list? I don't believe it should as llvm::MemoryBuffer doesn't keep the fd around after it has mmapped a file.

**Comment by klausler** [comment]
> @sscalpone could you check if this has the defective fd behaviour you mentioned on the mailing list? I don't believe it should as llvm::MemoryBuffer doesn't keep the fd around after it has mmapped a file.

If the open file descriptor is closed, then the file is not mapped into the virtual address space.

**Comment by DavidTruby** [comment]
> If the open file descriptor is closed, then the file is not mapped into the virtual address space.

Per the POSIX standard, this isn't the case. See https://pubs.opengroup.org/onlinepubs/7908799/xsh/mmap.html:
" The mmap() function adds an extra reference to the file associated with the file descriptor fildes which is not removed by a subsequent close() on that file descriptor. This reference is removed when there are no more mappings to the file. "

File: lib/Parser/CMakeLists.txt [hunk 1]
```diff
@@ -33,7 +33,7 @@ add_library(FortranParser
 )
 
 target_link_libraries(FortranParser
-  FortranCommon
+  FortranCommon LLVMSupport
```
Comment (kiranchandramohan) [inline lib/Parser/CMakeLists.txt]: General style seems to be one entry per library/file.

**Comment by klausler** [comment]
> > If the open file descriptor is closed, then the file is not mapped into the virtual address space.
> 
> Per the POSIX standard, this isn't the case. See https://pubs.opengroup.org/onlinepubs/7908799/xsh/mmap.html:
> " The mmap() function adds an extra reference to the file associated with the file descriptor fildes which is not removed by a subsequent close() on that file descriptor. This reference is removed when there are no more mappings to the file. "

So the limit on the number of simultaneous open file descriptors won't be a problem?  That's my only concern here.  Have you tested this with a large number of `INCLUDE` files?

**Comment by DavidTruby** [comment]
> So the limit on the number of simultaneous open file descriptors won't be a problem? That's my only concern here. Have you tested this with a large number of INCLUDE files?

It shouldn't be a problem, as we shouldn't have that many fds open at any one time. I don't have a test case that would have enough files to reach the fd limit though.

Clang and Swift both use this MemoryBuffer class for their source file managing so I assume this has already been considered. We should check it though, I'll see if I can randomly generate a case.

**Comment by DavidTruby** [comment]
> We should check it though, I'll see if I can randomly generate a case.

I've tried it with a file that #includes 10,000 other files, which is well over my fd limit per process on my system. I don't get any errors and get correct output from `-fget-symbols-sources`.

File: lib/Parser/CMakeLists.txt (hunk 1 above)
Comment (DavidTruby) [reply lib/Parser/CMakeLists.txt]: Done!

File: lib/Parser/parsing.cpp [hunk 2]
```diff
@@ -29,12 +29,13 @@ const SourceFile *Parsing::Prescan(const std::string &path, Options options) {
     }
   }
 
-  std::stringstream fileError;
+  std::string fileError_buf;
+  llvm::raw_string_ostream fileError{fileError_buf};
```
Comment (tskeith) [inline lib/Parser/parsing.cpp]: How is llvm::raw_string_ostream better than std::stringstream?

File: lib/Parser/preprocessor.cpp [hunk 3]
```diff
@@ -574,8 +574,9 @@ void Preprocessor::Directive(const TokenSequence &dir, Prescanner *prescanner) {
           "#include: empty include file name"_err_en_US);
       return;
     }
-    std::stringstream error;
-    const SourceFile *included{allSources_.Open(include, &error)};
+    std::string error_buf;
+    llvm::raw_string_ostream error{error_buf};
```
Comment (tskeith) [inline lib/Parser/preprocessor.cpp]:
```suggestion
    std::string errorBuf;
    llvm::raw_string_ostream error{errorBuf};
```

File: lib/Parser/source.cpp [hunk 4]
```diff
@@ -82,183 +58,68 @@ std::string LocateSourceFile(
   }
   for (const std::string &dir : searchPath) {
     std::string path{dir + '/' + name};
-    struct stat statbuf;
-    if (stat(path.c_str(), &statbuf) == 0 && !S_ISDIR(statbuf.st_mode)) {
+    bool is_dir = false;
+    auto er = llvm::sys::fs::is_directory(path, is_dir);
+    if (!er && !is_dir) {
```
Comment (tskeith) [inline lib/Parser/source.cpp]:
```suggestion
    bool isDir{false};
    auto er{llvm::sys::fs::is_directory(path, isDir)};
    if (!er && !isDir) {
```

File: lib/Parser/parsing.cpp (hunk 2 above)
Comment (DavidTruby) [reply lib/Parser/parsing.cpp]: Use of llvm::raw_ostream over std::ostream is [mandated for new code going in to llvm](https://llvm.org/docs/CodingStandards.html#use-raw-ostream) and this has been bought up on the mailing list with respect to f18, so if our goal is to submit to llvm need to change over.

Comment (tskeith) [reply lib/Parser/parsing.cpp]: There is no good reason not to use std::stringstream and it's not clear that's what that link says. Right above there it says:
> Note that using the other stream headers (`<sstream>` for example) is not problematic in this regard

Comment (DavidTruby) [reply lib/Parser/parsing.cpp]: It does say that sstream is not problematic in the regard that it doesn't introduce global objects with non-static constructors. However, it goes on to say:
> New code should always use raw_ostream for writing

Which I think is fairly clear, and we certainly are new code from LLVM's perspective.

Comment (sscalpone) [reply lib/Parser/parsing.cpp]: I believe the context of the quote "New code should always use...." pertains to writing files.  In this case, stringstream seems better and appropriate.

File: lib/Parser/source.cpp [hunk 5]
```diff
@@ -82,183 +58,68 @@ std::string LocateSourceFile(
...
-    }
-    std::size_t chunk = crcp - p;
-    std::memmove(buffer + wrote, p, chunk);
-    wrote += chunk;
-    p += chunk + 1;
-    bytes -= chunk + 1;
-  }
-  return wrote;
+std::size_t RemoveCarriageReturns(llvm::MutableArrayRef<char> buf) {
+  auto end = llvm::remove_if(buf, [](const char c) { return c == '\r'; });
```
Comment (sscalpone) [inline lib/Parser/source.cpp]: Do you know if the lambda is inlined?  If not, we should probably check the performance vs the original loop & memmove.

File: lib/Semantics/mod-file.cpp [hunk 6]
```diff
@@ -719,8 +719,8 @@ static std::string CheckSum(const std::string_view &contents) {
   return result;
 }
 
-static bool VerifyHeader(const char *content, std::size_t len) {
-  std::string_view sv{content, len};
+static bool VerifyHeader(llvm::ArrayRef<char> content) {
```
Comment (sscalpone) [inline lib/Semantics/mod-file.cpp]: Why not just pass the sourceFile?

File: lib/Parser/source.cpp [hunk 7]
```diff
@@ -10,63 +10,39 @@
...
-    static constexpr int BOMBytes{3};
-    static const char UTF8_BOM[]{"\xef\xbb\xbf"};
-    if (bytes_ >= BOMBytes && std::memcmp(content_, UTF8_BOM, BOMBytes) == 0) {
-      content_ += BOMBytes;
-      bytes_ -= BOMBytes;
-      encoding_ = Encoding::UTF_8;
-    }
+  llvm::StringRef content{buf_->getBufferStart(), buf_->getBufferSize()};
+  constexpr llvm::StringLiteral UTF8_BOM{"\xef\xbb\xbf"};
+  if (content.startswith(UTF8_BOM)) {
```
Comment (sscalpone) [inline lib/Parser/source.cpp]: See how content_ is bumped up in the original. Does that happen now?  How are those BOM characters handled?

File: lib/Parser/source.cpp (hunk 5 above)
Comment (DavidTruby) [reply lib/Parser/source.cpp]: The original loop actually has O(n^2) complexity if I am not reading it wrong, whereas the complexity of remove_if is O(n). Regarding the specific question though, the lambda does get inlined by both compilers I've tested it on.

File: lib/Parser/source.cpp (hunk 7 above)
Comment (DavidTruby) [reply lib/Parser/source.cpp]: The `bom_end_` member variable is set, and when `content()` is called it returns an ArrayRef starting after that end. So anybody that calls `content()` will not see the BOM.

File: lib/Semantics/mod-file.cpp (hunk 6 above)
Comment (DavidTruby) [reply lib/Semantics/mod-file.cpp]: Because we need to ignore the BOM and the carraige returns, so we need the slicing behaviour of `content()`

File: lib/Parser/parsing.cpp (hunk 2 above)
Comment (DavidTruby) [reply lib/Parser/parsing.cpp]: I believe this was specifically clarified by Hal on a community call a few weeks ago, that the policy is that the use of any of the stream libraries including `std::stringstream` is highly discouraged for new code.

Could you explain why stringstream is better and more appropriate here? `std::stringstream` and `llvm::raw_string_ostream` have the same interface.

Comment (tskeith) [reply lib/Parser/parsing.cpp]: > Could you explain why stringstream is better and more appropriate here?

It's part of the language.

> `std::stringstream` and `llvm::raw_string_ostream` have the same interface.

This change shows a case where that's not the case.

Comment (DavidTruby) [reply lib/Parser/parsing.cpp]: > Could you explain why stringstream is better and more appropriate here?

And `llvm::raw_string_ostream` is a part of the library of the project we are supposed to be a part of, and is preferred by that project for a number of technical reasons that are outlined in the documentation. I fail to see how this makes `std::stringstream` better.

> This change shows a case where that's not the case.

Are you referring to the fact that a separate buffer needs to be stored here? That is a minor change in the construction of the class but the observable interface as used here is identical.

File: lib/Parser/source.cpp (hunk 7 above)
Comment (DavidTruby) [reply lib/Parser/source.cpp]: Everything outside this class should only be able to see the `content()` and `bytes()` functions that correctly deal with the bom and the carriage returns. So the observable behaviour of this class should be no different to before, where the content and number of bytes were adjusted. There wasn't before and isn't now a way to change the actual size of the mapped buffer without writing to a file and remapping after dealing with the bom and carriage returns.

Effectively what I'm saying is I don't know the exact answer wrt LSP, but if it worked before it will work now (modulo bugs of course)

File: lib/Parser/parsing.cpp (hunk 2 above)
Comment (RichBarton-Arm) [reply lib/Parser/parsing.cpp]: We discussed this on the flang technical call on March 9 and agreed:
 -  The LLVM Coding Guidelines are unclear on whether or not sstream should be allowed
 -  The long-time LLVM community folks felt that the intention is for sstream not to be used at all in preference to LLVM APIs.
 - We would make the change that David suggests in F18
 - Johannes would start a thread to clarify the wording in the LLVM Coding Standards to make clear the intent.

Comment (Meinersbur) [reply lib/Parser/parsing.cpp]: The use of `std::stringstream` in Polly was removed: https://github.com/llvm/llvm-project/commit/0e93f3b0a06fc470e37b5be99734f5105ba11232.

**Comment by DavidTruby** [comment]
@sscalpone @tskeith is this ready to merge now?

File: lib/Parser/source.cpp [hunk 8]
```diff
@@ -82,183 +60,68 @@ std::string LocateSourceFile(
...
+bool SourceFile::Open(std::string path, llvm::raw_ostream &error) {
   Close();
   path_ = path;
-  std::string errorPath{"'"s + path + "'"};
-  errno = 0;
-  fileDescriptor_ = open(path.c_str(), O_RDONLY);
-  if (fileDescriptor_ < 0) {
-    *error << "Could not open " << errorPath << ": " << std::strerror(errno);
+  std::string errorPath{"'"s + path_ + "'"};
+  auto buf_or = llvm::WritableMemoryBuffer::getFile(path);
```
Comment (tskeith) [inline lib/Parser/source.cpp]: Please use camel case and braced initialization.

File: lib/Parser/source.cpp (hunk 5 above)
Comment (klausler) [reply lib/Parser/source.cpp]: Why do you think that the loop is O(n**2)?  It makes one pass over the source, one iteration per `\r`, and moves each of the other bytes exactly once.  And the new version doesn't use `memchr`, which has been highly optimized for SIMD.  I'd like to see actual measured performance comparisons before signing off on this.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: Memmove is an O(n) function and is called inside an O(n) loop. In the worst case where every character is a carriage return n^2 operations will have been performed.

Comment (klausler) [reply lib/Parser/source.cpp]: O(#of lines) * O(average line length) == O(#of lines) * O(#total bytes / #lines) == O(#total bytes).

Comment (DavidTruby) [reply lib/Parser/source.cpp]: Sorry, I don't follow. Consider the pathological case where every character in the input is a carriage return. We then call memmove once per character (which is n characters), and each memmove is O(n).

Comment (klausler) [reply lib/Parser/source.cpp]: And each of those `memmove()` calls will be moving zero bytes.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: If the memmove moves 0 bytes then the array remains the same and we haven't removed any carriage returns?

Comment (DavidTruby) [reply lib/Parser/source.cpp]: If the memmove moves 0 bytes then the array remains the same and we haven't removed any carriage returns?

Comment (klausler) [reply lib/Parser/source.cpp]: The `memmove()` calls are moving the chunks of source bytes that lie *between* the `\r` characters.  If the whole buffer contains nothing but `\r` characters, those chunks are all empty.  The final size of the compacted buffer is the sum of the sizes of these chunks, so that would be zero.

Comment (klausler) [reply lib/Parser/source.cpp]: To add some data to the discussion:  here's a simple timing harness for testing three alternative implementations on a megabyte-sized source buffer with `\r` characters every 20 bytes (a guess at a reasonable mean line length, change it if you like).  I compiled this with g++ 9.2.0 and ran it on an old Haswell deskside machine with its Linux clock governor set to the `performance` mode.  The overhead of filling the source array 10,000 times is 0.86 total seconds.  The time needed to fill the source buffer and run the existing `RemoveCarriageReturns()` routine 10,000 times was 4.34 seconds.  The time for the `ByteAtATime()` alternative was 6.66 seconds.  The time for the predicate-based `remove_if()` alternative was 21.62 seconds.

```
#include <cstdlib>
#include <cstring>
#include <functional>
#include <iostream>

static std::size_t RemoveCarriageReturns(char *buffer, std::size_t bytes) {
  std::size_t wrote{0};
  char *p{buffer};
  while (bytes > 0) {
    void *vp{static_cast<void *>(p)};
    void *crvp{std::memchr(vp, '\r', bytes)};
    char *crcp{static_cast<char *>(crvp)};
    if (!crcp) {
      std::memmove(buffer + wrote, p, bytes);
      wrote += bytes;
      break;
    }
    std::size_t chunk = crcp - p;
    std::memmove(buffer + wrote, p, chunk);
    wrote += chunk;
    p += chunk + 1;
    bytes -= chunk + 1;
  }
  return wrote;
}

static std::size_t ByteAtATime(char *buffer, std::size_t bytes) {
  char *q{buffer};
  for (char *p{buffer}; bytes-- > 0; ++p) {
    if (*p != '\r') {
      *q++ = *p;
    }
  }
  return q - buffer;
}

static std::size_t remove_if(char *buffer, std::size_t bytes, const std::function<bool(char)> &predicate) {
  char *q{buffer};
  for (char *p{buffer}; bytes-- > 0; ++p) {
    if (!predicate(*p)) {
      *q++ = *p;
    }
  }
  return q - buffer;
}

int main() {
  int n{1 << 20};
  int linelen{20};
  char *src{static_cast<char *>(std::malloc(n))};
  int left;
  for (int reps{0}; reps < 10000; ++reps) {
    std::memset(src, 0, n);
    for (int j{0}; j < n; j += linelen) {
      src[j] = '\r';
    }
// Uncomment one of these three lines
//    int left = RemoveCarriageReturns(src, n);
//    int left = ByteAtATime(src, n);
//    int left = remove_if(src, n, [](char c){ return c == '\r'; });
    if (reps == 0) {
      std::cout << left << '\n';
    }
  }
  return 0;
}
```

Comment (DavidTruby) [reply lib/Parser/source.cpp]: In order to find the next non - carriage return you have to have called another O(n) function though? 

Regardless, I was asked specifically by someone on llvm-dev to change the way character operations were done in this file to use C++ algorithms, which is why I made the change. On my system this function vectorises fine.
 I won't have time to do a thorough performance analysis before the merge and take that to llvm-dev to get consensus that the change doesn't need to be made. If we still think this is necessary can we do it as a post merge review? Otherwise we will need to delay the merge with llvm.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: Why have you written your own remove_if function? The one here implies 2 pointer indirections per predicate call (due to how std::function is implemented) and isn't inlineable. Neither of these things are true of std::remove_if. Your library implementation possibly also includes vectorisation hints

Comment (klausler) [reply lib/Parser/source.cpp]: Yes, I'm calling `memchr()` and `memmove()` in the loop.  Those are indeed O(n) functions.  But that does *not* make the overall function O(n**2), nor does it change the actual measured advantages of using optimized library routines in cases where performance matters.

I'm done here.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: std::remove_if is _also_ an optimised library routine, and you haven't measured the performance of that that I can see. Since that's the one I'm calling that's really the measurement we need to consider isn't it?

Comment (klausler) [reply lib/Parser/source.cpp]: Using `std::remove_if()` yields a time of 11.81 seconds, so it's about half the speed of the `ByteAtATime()` routine or a third the speed of the optimized routine.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: This change was as a prerequisite of the merge in this thread on llvm-dev: http://lists.llvm.org/pipermail/llvm-dev/2020-February/139458.html
If we aren't going to make the change for the reasons you've outlined would you possibly be able to take that discussion up on the list?

Comment (klausler) [reply lib/Parser/source.cpp]: I don't subscribe to any of these mailing lists; sorry.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: I can't replicate your performance results above; I've put my findings using each compiler and standard library combination I regularly test with below. 

I have only done a single run each time so the margin of error is larger than I would like; but in essence if we assume 0.5s to be our margin of error (which seems reasonable given the below results) then there is only one case when RemoveCarriageReturns is faster than std::remove_if, which seems to be a regression in libcxx9. 

On x86 macOS std::remove_if gives slightly faster performance than RemoveCarriageReturns, albeit not hugely outside my stated margin of error, and on Arm64 std::remove_if is around _40% faster_ than RemoveCarriageReturns when using libstdcxx, and within the margin of error when using libcxx.

These results were obtained using your test program with the additional `remove_if` line you gave, and in each case I only used the flags `-std=c++17` and `-O3`. A quick test of a couple of them with `-O2` demonstrated that there was no difference between the two optimisation levels, so I continued to use the one that LLVM and the reworked CMake for LLVM use.

# Arm64 Linux
## Clang 9 libstdcxx 9
`std::remove_if` - 0m7.606s
`RemoveCarriageReturns` - 0m12.148s

## Clang 9 libcxx 9
`std::remove_if` - 0m14.480s
`RemoveCarriageReturns` - 0m12.239s

## Clang 8 libstdcxx 9
`std::remove_if` - 0m7.563s
`RemoveCarriageReturns` - 0m12.219s

## Clang 8 libcxx 8
`std::remove_if` - 0m12.134s
`RemoveCarriageReturns` - 0m12.103s

## Clang 7 libstdcxx 9
`std::remove_if` - 0m7.504s
`RemoveCarriageReturns` - 0m12.097s

## GCC 9
`std::remove_if` - 0m6.894s
`RemoveCarriageReturns` - 0m12.701s

## GCC 8
`std::remove_if` - 0m6.955s
`RemoveCarriageReturns` - 0m12.646s

## GCC 7
`std::remove_if` - 0m6.946s
`RemoveCarriageReturns` - 0m12.585s

# x86-64 macOS
## Clang 9 libcxx 9
`std::remove_if` - 0m3.316s
`RemoveCarriageReturns` - 0m4.629

## Clang 8 libcxx 8
`std::remove_if` - 0m3.641s
`RemoveCarriageReturns` - 0m4.873s

## Clang 7 libcxx 7
`std::remove_if` - 0m4.345s
`RemoveCarriageReturns` - 0m4.836s

## Apple Clang 11
`std::remove_if` - 0m4.454s
`RemoveCarriageReturns` - 0m4.413s

Comment (klausler) [reply lib/Parser/source.cpp]: And were you able to reproduce my results (gcc 9.2.0, -O3, x86, cpufreq governor set to `performance`)?

Comment (klausler) [reply lib/Parser/source.cpp]: What were the times from the `ByteAtATime()` alternative in these environments?

Comment (DavidTruby) [reply lib/Parser/source.cpp]: I cannot replicate your results on X86 Linux with gcc and performance governor either, see below.
I didn't test the other implementation as it isn't being proposed here.

# x86_64 Linux
## GCC 9
`std::remove_if` - 0m3.363s
`RemoveCarriageReturns` - 0m4.775s

## GCC 8
`std::remove_if` - 0m3.388s
`RemoveCarriageReturns` - 0m7.851s

## GCC 7
`std::remove_if` - 0m5.688s
`RemoveCarriageReturns` - 0m8.083s

Comment (klausler) [reply lib/Parser/source.cpp]: The reason to measure the `ByteAtAtTime()` alternative is that it's a sanity check on the relationship between the other measurements.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: I have added results for `ByteAtATime` above

Comment (DavidTruby) [reply lib/Parser/source.cpp]: As a sanity check I think that ByteAtATime looks odd in your results; ByteAtATime is effectively the “simple” implementation of std::remove_if, if that function had no optimisations in the library. So remove_if should be at least as fast as ByteAtATime, as the performance of ByteAtATime is what remove_if would have if it had no optimisations in the library implementation.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: I have managed to somewhat replicate your results on my home desktop. Using a timer that only measures the loops, increasing the number of iterations to get clearer results and with the times given in microseconds I get;
`std::remove_if` - 698210 average per iteration
`RemoveCarriageReturns` - 503866
`ByteAtATime` - 693623

So I don't see any significant difference between `ByteAtATime` and `std::remove_if` still, but `RemoveCarriageReturns` is faster on this specific setup.

Comment (RichBarton-Arm) [reply lib/Parser/source.cpp]: How can we move forward here?

Clearly we don't want to degrade performance unnecessarily, but that is not the only concern. The coding style of the proposed new implementation is certainly closer to what the LLVM community would expect and this particular code has been requested to be re-written - http://lists.llvm.org/pipermail/llvm-dev/2020-February/139464.html - before we upstream.

Given the data shown, I don't think it is clear that David's proposed new implementation is systematically worse than the current implementation. It seems that results vary depending probably on how optimised your C and C++ standard libraries are for the system you are on,

If we agree on that, I think we should make the proposed change on the grounds of coding style alignment to LLVM to unblock upstreaming.

Comment (klausler) [reply lib/Parser/source.cpp]: Dealing with DOS line endings is something that we have to do most often on Windows than elsewhere, so the x86 timings seem to indicate to me that we'll want to retain the fast approach in NVIDIA's product sources.  You can do whatever you think you have to in your LLVM fork.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: I think in order to make any statements about windows we would have to test there rather than Linux; their C and C++ standard library implementations are completely different to the ones on Linux and therefore very likely to have different performance characteristics. In addition my macOS results still show `remove_if` as being faster than the hand rolled function here, and that's also on x86. Again I suspect that this is because the C standard library on macOS is different to on Linux.

Comment (DavidTruby) [reply lib/Parser/source.cpp]: This function does also have to run on every platform, regardless of whether the input actually has carriage returns or not. So the performance matters everywhere not just on windows on x86.

Comment (RichBarton-Arm) [reply lib/Parser/source.cpp]: @sscalpone 
So that we can progress the rest of the patch, David has removed this change from the review. Can we merge as-is?

**Comment by DavidTruby** [comment]
@sscalpone rebased on top of the LLVM streams patch

**Comment by sscalpone** [comment]
@DavidTruby This PR causes a fatal check when compiling Nyx/Exec/Scaling in my sandbox. I don't know if the problem is with the PR or if the PR is exposing a different issue.

`fatal internal error: CHECK(range_.Contains(at)) failed at /home/sjs/work/pgi/f18/pr1032/f18/lib/Parser/provenance.cpp(389)`

I can't share the Nyx source tree that I'm using; however, I think it is based on this:
`https://github.com/AMReX-Astro/Nyx/tree/master/Exec/Scaling`.

**Comment by DavidTruby** [comment]
> @DavidTruby This PR causes a fatal check when compiling Nyx/Exec/Scaling in my sandbox. I don't know if the problem is with the PR or if the PR is exposing a different issue.
> 
> `fatal internal error: CHECK(range_.Contains(at)) failed at /home/sjs/work/pgi/f18/pr1032/f18/lib/Parser/provenance.cpp(389)`
> 
> I can't share the Nyx source tree that I'm using; however, I think it is based on this:
> `https://github.com/AMReX-Astro/Nyx/tree/master/Exec/Scaling`.
> 
> The failing file is zero length.
> 
> ```
> % pwd
> .../pr1032/build
> % touch zero.f90
> % ls -al zero.f90
> -rw-rw-r-- 1 xxx yy 0 Mar 22 08:46 zero.f90
> % tools/f18/bin/f18 zero.f90
> 
> fatal internal error: CHECK(range_.Contains(at)) failed at /home/sjs/work/pgi/f18/pr1032/f18/lib/Parser/provenance.cpp(389)
> Aborted
> ```

@sscalpone I seem to be able to reproduce this with empty files. I'll fix it and add a lit test for empty files.

**Comment by DavidTruby** [comment]
@sscalpone Should be fixed now. I've left it as a separate commit so you can review the fix, please let me know if it's ok to squash.

**Comment by DavidTruby** [comment]
@sscalpone I've squashed this to one commit and written a more descriptive error message
